### PR TITLE
Add Unix daemon mode, syslog and journald support, more signal handling

### PR DESCRIPTION
**Pull request by bluca**
This is a bigger one - the main new features are support for new command line flags to behave like a traditional Unix daemon (fork in the background, record PID file, change user/group, stop output to stdout/stderr) and a refactor in the log (message) handling class to add support for writing to syslog and/or to systemd's journald (with disabled by default optional build-dependency on the relevant library via CMake).

The logging levels have been matched to the standard syslog ones by mapping, to avoid having to chang every single source file. They more or less matched already so it was easy.

Something that we found extremely useful is being able to change the log level at runtime, without restarting. A common way to achieve that is via SIGUSR1/2 to go up/down, and that's what I implemented.
NOTE: in order to do that, it was required to move the reload signal to SIGHUP, like it is used traditionally.
Given reloading didn't really support reusing modules until one of my recent PRs, I hope this is not a prob lem.

But given it is a backward incompatible change I am totally fine if you would like me to make it optional. Perhaps a build-time CMake switch, or a runtime option. Let me know.

### CODE REVIEW

File: src/common/msg.cc [hunk 1]
```diff
@@ -95,15 +95,26 @@ extern "C" {
...
 	void msg_shutdown()
 	{
-		pthread_mutex_destroy(&msg_mutex);
+		if (msg_get_syslog()) {
+			closelog();
+		}
+
+		int retval = pthread_mutex_destroy(&msg_mutex);
+		if (retval != 0) {
+			printf("!!! msg: pthread_mutex_destroy returned error code %d (%s)\n", retval, strerror(retval));
```
Comment (ogasser) [inline src/common/msg.cc]: Should we use `fprintf(stderr,…)` instead of `printf(…)` for error messages?

File: src/common/msg.cc [hunk 2]
```diff
@@ -120,7 +131,9 @@ extern "C" {
 		// we must lock via mutex, else logging outputs are mixed when several
 		// threads log simultaneously
 		int retval = pthread_mutex_lock(&msg_mutex);
-		if (retval != 0) printf("!!! msg: pthread_mutex_lock returned error code %d (%s)\n", retval, strerror(retval));
+		if (retval != 0) {
+			printf("!!! msg: pthread_mutex_lock returned error code %d (%s)\n", retval, strerror(retval));
```
Comment (ogasser) [inline src/common/msg.cc]: Should we use `fprintf(stderr,…)` instead of `printf(…)` for error messages?

File: src/common/msg.cc [hunk 3]
```diff
@@ -158,7 +171,9 @@ extern "C" {
 			vsnprintf(logtext, EXCEPTION_MAXLEN-strlen(logtext), fmt, *args);
 		}
 		retval = pthread_mutex_unlock(&msg_mutex);
-		if (retval != 0) printf("!!! msg: pthread_mutex_unlock returned error code %d\n", retval);
+		if (retval != 0) {
+			printf("!!! msg: pthread_mutex_unlock returned error code %d\n", retval);
```
Comment (ogasser) [inline src/common/msg.cc]: Should we use `fprintf(stderr,…)` instead of `printf(…)` for error messages?

File: src/common/msg.cc (hunk 1 above)
Comment (bluca) [inline src/common/msg.cc]: Sure, done

File: src/vermont.cc [hunk 4]
```diff
@@ -139,10 +160,25 @@ usage (int status)
...
+			"                                times and mix-matched. \n"
+			"                                In increasing order:\n\n"
+			"                                    debug\n"
+			"                                    info\n"
+			"                                    notice\n"
+			"                                    warning\n"
+			"                                    error\n"
+			"                                    critical\n\n"
+			"                                Default: critical, warning, error\n"
+			" -q, --quiet                Do not write output to console\n"
```
Comment (bluca) [inline src/vermont.cc]: done

File: src/vermont.cc [hunk 5]
```diff
@@ -42,9 +46,184 @@
...
+	bool daemon_mode;
+};
 
 
+static void
+record_pid (const char *pidfile)
+{
+	FILE *f = fopen(pidfile, "w");
+	if (!f) {
+		fprintf(stderr, "Could not open pid-file %s for writing", pidfile);
```
Comment (bluca) [inline src/vermont.cc]: Yes, done for this and all the others

File: src/vermont.cc [hunk 6]
```diff
@@ -167,7 +167,7 @@ usage (int status)
 			"                                    error\n"
 			"                                    critical\n\n"
 			"                                Default: critical, warning, error\n"
-			" -q, --quiet                Do not write output to console\n"
+			" -q, --quiet                Do not write output to console (implied by -b)\n"
 			" -b, --daemon               Run in daemon mode (implies -q)\n"
 			" -p, --pid-file FILE        Set process id filename (use with -d)\n"
```
Comment (ogasser) [inline src/vermont.cc]: Should be `(use with -b)` instead of `(use with -d)` for the below three lines?

Comment (bluca) [inline src/vermont.cc]: Ah good spot, fixed

**Comment by ogasser** [comment]
LGTM, thanks for the fixes!

**Comment by bluca** [comment]
Thanks! I just opened a PR with fixes for the unit test and for a lot of compiler warnings: https://github.com/tumi8/vermont/pull/59

**Pull request merged**